File: frontend/src/types/realEstate.ts
export interface RealEstateProperty {
  id: number

  name: string
  address: string | null
  property_type: string

  purchase_price: number
  current_value: number
  down_payment: number

  loan_original_amount: number
  loan_balance: number
  interest_rate: number | null

  monthly_payment: number
  monthly_rent: number
  monthly_expenses: number

  currency: string
  purchase_date: string | null
  loan_end_date: string | null

  equity: number

  annual_rent: number
  annual_expenses: number
  annual_loan_payments: number

  monthly_cash_flow: number
  annual_cash_flow: number

  gross_rental_yield: number
  net_rental_yield: number
  loan_to_value: number
}


export interface RealEstateSummary {
  properties_count: number

  total_current_value: number
  total_loan_balance: number
  total_equity: number

  total_monthly_rent: number
  total_monthly_expenses: number
  total_monthly_loan_payments: number

  total_monthly_cash_flow: number
  total_annual_cash_flow: number

  currency: string
}


export interface RealEstateCreate {
  name: string
  address: string | null
  property_type: string

  purchase_price: number
  current_value: number
  down_payment: number

  loan_original_amount: number
  loan_balance: number
  interest_rate: number | null

  monthly_payment: number
  monthly_rent: number
  monthly_expenses: number

  currency: string
  purchase_date: string | null
  loan_end_date: string | null
}


export type RealEstateUpdate = Partial<
  RealEstateCreate
>


export interface RealEstateProjectionPoint {
  year: number
  projection_date: string

  property_value: number
  loan_balance: number
  equity: number

  principal_paid: number
  interest_paid: number
}


export interface RealEstateProjection {
  property_id: number
  property_name: string
  currency: string

  annual_property_growth: number
  annual_interest_rate: number
  monthly_payment: number

  loan_end_date: string | null
  projection_years: number

  total_principal_paid: number
  total_interest_paid: number

  points: RealEstateProjectionPoint[]
}

File: backend/schemas/real_estate.py
from datetime import date

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
)


class RealEstateBase(BaseModel):
    name: str = Field(
        min_length=1,
        max_length=100,
    )

    address: str | None = Field(
        default=None,
        max_length=250,
    )

    property_type: str = Field(
        default="Apartment",
        min_length=1,
        max_length=50,
    )

    purchase_price: float = Field(
        ge=0,
    )

    current_value: float = Field(
        ge=0,
    )

    down_payment: float = Field(
        default=0.0,
        ge=0,
    )

    loan_original_amount: float = Field(
        default=0.0,
        ge=0,
    )

    loan_balance: float = Field(
        default=0.0,
        ge=0,
    )

    interest_rate: float | None = Field(
        default=None,
        ge=0,
    )

    monthly_payment: float = Field(
        default=0.0,
        ge=0,
    )

    monthly_rent: float = Field(
        default=0.0,
        ge=0,
    )

    monthly_expenses: float = Field(
        default=0.0,
        ge=0,
    )

    currency: str = Field(
        default="EUR",
        min_length=3,
        max_length=10,
    )

    purchase_date: date | None = None

    loan_end_date: date | None = None


class RealEstateCreate(
    RealEstateBase
):
    pass


class RealEstateUpdate(BaseModel):
    name: str | None = Field(
        default=None,
        min_length=1,
        max_length=100,
    )

    address: str | None = Field(
        default=None,
        max_length=250,
    )

    property_type: str | None = Field(
        default=None,
        min_length=1,
        max_length=50,
    )

    purchase_price: float | None = Field(
        default=None,
        ge=0,
    )

    current_value: float | None = Field(
        default=None,
        ge=0,
    )

    down_payment: float | None = Field(
        default=None,
        ge=0,
    )

    loan_original_amount: float | None = Field(
        default=None,
        ge=0,
    )

    loan_balance: float | None = Field(
        default=None,
        ge=0,
    )

    interest_rate: float | None = Field(
        default=None,
        ge=0,
    )

    monthly_payment: float | None = Field(
        default=None,
        ge=0,
    )

    monthly_rent: float | None = Field(
        default=None,
        ge=0,
    )

    monthly_expenses: float | None = Field(
        default=None,
        ge=0,
    )

    currency: str | None = Field(
        default=None,
        min_length=3,
        max_length=10,
    )

    purchase_date: date | None = None

    loan_end_date: date | None = None


class RealEstateResponse(
    RealEstateBase
):
    id: int

    equity: float
    annual_rent: float
    annual_expenses: float
    annual_loan_payments: float
    monthly_cash_flow: float
    annual_cash_flow: float
    gross_rental_yield: float
    net_rental_yield: float
    loan_to_value: float

    model_config = ConfigDict(
        from_attributes=True,
    )


class RealEstateSummary(BaseModel):
    properties_count: int
    total_current_value: float
    total_loan_balance: float
    total_equity: float
    total_monthly_rent: float
    total_monthly_expenses: float
    total_monthly_loan_payments: float
    total_monthly_cash_flow: float
    total_annual_cash_flow: float
    currency: str